client/shared/src/main/scala/org/http4s/client/Client.scala

```diff
@@ -289,16 +291,59 @@ object Client {
       go(source).stream
     }
 
-    def runOrDispose(req: Request[F]): F[Resource[F, Response[F]]] =
-      Ref[F].of(false).map { disposed =>
-        val req0 = addHostHeaderIfUriIsAbsolute(req.pipeBodyThrough(until(disposed)))
-        Resource
-          .eval(app(req0))
-          .onFinalize(disposed.set(true))
-          .map(_.pipeBodyThrough(until(disposed)))
-      }
+    def run(
+        message: Message[F],
+        refOp: Option[Ref[F, Boolean]] = None,
+    ): Resource[F, Response[F]] = message match {
+      case req: Request[F] =>
+        Resource.eval(Ref[F].of(false)).flatMap { disposed =>
+          val reqAugmented =
+            addHostHeaderIfUriIsAbsolute(req.pipeBodyThrough(until(disposed)))
+          Resource
+            .eval(app(reqAugmented))
+            .onFinalize(disposed.set(true))
+            .flatMap(run(_, Option(disposed)))
+        }
+      case resp: Response[F] =>
+        for {
+          disposed <- Resource.eval(refOp.fold(Ref[F].of(false))(F.pure))
+          channel <- Resource.eval(Channel.synchronous[F, Chunk[Byte]])
+
+          _ <- resp.body.chunks
+            .through(channel.sendAll)
+            .compile
+            .drain
+            .background
+
+          ifDisposed = Stream
+            .eval(disposed.get)
+            .ifM(
+              Stream.eval(F.raiseError(new IOException("response was disposed"))),
+              Stream.empty,
+            )
+
+          r = resp.withBodyStream(
+            channel.stream.unchunks
+              .ifEmpty(ifDisposed)
+              .onFinalize(
+                channel.stream.compile.drain
+                  .guarantee(disposed.set(true))
+              )
+          )
+
+          _ <- Resource.onFinalize {
+            disposed.get
+              .ifM(
+                F.unit,
+                r.body.compile.drain,
+              )
```

Comment:
Should this also `disposed.set(true)`? In case user tries to consume outside of `use` block.

Also, probably this doesn't need to be conditional, it should be safe to just drain again.

Reply:
It can be unconditional if we don't have to return the `disposed` exception.

Reply:
> I noticed that it calls the body stream which has this logic in the final condition, so it will be executed in any case. Additionally, it should be conditional; otherwise, the logic will trigger the disposed error check in the body stream.

Oh, sorry, I missed that! Right, you are calling `r.body.compile.drain`, not `channel.stream.compile.drain`. I think this is fine then 👍